go-controller/pkg/ovn/egressfirewall_resolver.go

```diff
@@ -0,0 +1,139 @@
+	}
+}
+
+// Add is called whenever a DNS name is needed to be added or updated.
+func (res *Resolver) Add(dnsName string, addresses []string, updateAddressSet bool) (addressset.AddressSet, error) {
+	res.dnsLock.Lock()
+	defer res.dnsLock.Unlock()
+
+	// Get the details of the resolved name corresponding to the DNS name.
+	resolvedName, exists := res.dnsNames[dnsName]
+
+	// If the details of the resolved name is not available, then create
+	// the address set for the DNS name.
+	if !exists {
+		if res.addressSetFactory == nil {
+			return nil, fmt.Errorf("error adding DNS name %s: addressSetFactory is nil", dnsName)
+		}
+		asIndex := getEgressFirewallDNSAddrSetDbIDs(dnsName, res.controllerName)
+		dnsAddressSet, err := res.addressSetFactory.NewAddressSet(asIndex, nil)
+		if err != nil {
+			return nil, fmt.Errorf("cannot create AddressSet for DNS name %s: %v", dnsName, err)
+		}
+		resolvedName.dnsAddressSet = dnsAddressSet
+	}
+
+	// If updateAddressSet is set to true then update the address set of
+	// the DNS name with the IPs. The IPs which don't match the ip mode
+	// and those which match the cluster subnet are ignored.
+	if updateAddressSet {
+		var ips []net.IP
+		for _, addr := range addresses {
+			ip := net.ParseIP(addr)
+			if ip != nil {
+				ips = append(ips, ip)
+			}
+		}
+		resolvedName.ips = ips
+
+		// Ignore ips which don't match the configured ip mode.
+		// Ignore ips from clusterSubnet, since this subnet shouldn't be affected by egress firewall.
+		filteredIPs := []net.IP{}
+		for _, ip := range ips {
+			ignoreIP := false
+
+			if config.IPv4Mode && !config.IPv6Mode && ip.To4() == nil {
+				ignoreIP = true
+			} else if !config.IPv4Mode && config.IPv6Mode && ip.To4() != nil {
+				ignoreIP = true
+			}
+
+			if !ignoreIP {
+				for _, clusterSubnet := range config.Default.ClusterSubnets {
+					if clusterSubnet.CIDR.Contains(ip) {
+						ignoreIP = true
+						break
+					}
+				}
+			}
+
+			if !ignoreIP {
+				filteredIPs = append(filteredIPs, ip)
+			}
+		}
+		if err := resolvedName.dnsAddressSet.SetIPs(filteredIPs); err != nil {
+			return nil, fmt.Errorf("cannot add IPs to AddressSet for DNS name %s: %v", dnsName, err)
+		}
+	}
+
+	// Add the updated resolved name to the dnsNames map corresponding to the DNS name.
+	res.dnsNames[dnsName] = resolvedName
+
+	return resolvedName.dnsAddressSet, nil
+}
+
+// Delete is called whenever a DNS name is needed to be deleted.
+func (res *Resolver) Delete(dnsName string) error {
+	res.dnsLock.Lock()
+	defer res.dnsLock.Unlock()
+
+	// Get the resolved name details corresponding to the DNS name.
+	resolvedName, exists := res.dnsNames[dnsName]
+	// Return if the corresponding resolved name details is not found.
+	if !exists {
+		klog.Errorf("Details missing for DNS name %s. Skipping address set destroy operation.", dnsName)
+		return nil
+	}
+
+	// Delete the address set associated with the DNS name.
+	err := resolvedName.dnsAddressSet.Destroy()
```

Comment:
this part may be tricky, because if we delete address set before the acls that reference it, ovn-controller will drop an error.
Leaving this comment to not forget, we can try to solve it on the next review iterations.

Reply:
I have modified this to check if the DNS name is still used in any namespace. When all the egress firewall objects are deleted it'll remove the corresponding namespaces and only then the address set can be destroyed. This will ensure that the ACLs are deleted first before the address sets are deleted.